lib/uplink/bucket.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"storj.io/storj/internal/memory"
+	"storj.io/storj/pkg/metainfo/kvmetainfo"
+	"storj.io/storj/pkg/storage/streams"
+	"storj.io/storj/pkg/storj"
+)
+
+// Bucket represents operations you can perform on a bucket
+type Bucket struct {
+	storj.Bucket
+
+	metainfo   *kvmetainfo.DB
+	streams    streams.Store
+	pathCipher storj.Cipher
+}
+
+// OpenObject returns an Object handle, if authorized.
+func (b *Bucket) OpenObject(ctx context.Context, path storj.Path) (o *Object, err error) {
+	defer mon.Task()(&ctx)(&err)
+	panic("TODO")
+}
+
+// UploadOptions controls options about uploading a new Object, if authorized.
+type UploadOptions struct {
+	// ContentType, if set, gives a MIME content-type for the Object.
+	ContentType string
+	// Metadata contains additional information about an Object. It can
+	// hold arbitrary textual fields and can be retrieved together with the
+	// Object. Field names can be at most 1024 bytes long. Field values are
+	// not individually limited in size, but the total of all metadata
+	// (fields and values) can not exceed 4 kiB.
+	Metadata map[string]string
+	// Expires is the time at which the new Object can expire (be deleted
+	// automatically from storage nodes).
+	Expires time.Time
+
+	// Volatile groups config values that are likely to change semantics
+	// or go away entirely between releases. Be careful when using them!
+	Volatile struct {
+		// DataCipher determines the ciphersuite to use for the Object's
+		// data encryption. If not set, the Bucket's default will be
+		// used.
+		DataCipher Cipher
+		// EncryptionBlockSize determines the unit size at which
+		// encryption is performed. See BucketConfig.EncryptionBlockSize
+		// for more information.
+		EncryptionBlockSize memory.Size
+
+		// RSParameters determines the Reed-Solomon and/or Forward Error
+		// Correction encoding parameters to be used for this Object.
+		RSParameters storj.RedundancyScheme
```

Comment:
Shouldn't we have a more generic name for this field? Theoretically `storj.RedundancyScheme` allows other encoding algorithms than Reed Solomon. This is why we came with `RedundancyScheme` as a more generic name for this type.

Reply:
That sounds fine. I didn't do that because I wanted to limit my changes, but you're right, it makes more sense.

Reply:
this is in the volatile section, right? but sure